include/deal.II/lac/qr.h

```diff
@@ -0,0 +1,863 @@
+ * This class and classes derived from it are meant to build $Q$ and $R$
+ * matrices one row/column at a time, i.e. by growing $R$ matrix from an empty
+ * $0\times0$ matrix to $N\timesN$, where $N$ is the number of added column
+ * vectors.
+ *
+ * As a consequence, matrices which have the same number of rows as each vector
+ * (i.e. $Q$ matrix) is stored as a collection of vectors of `VectorType`.
+ */
+template <typename VectorType>
+class BaseQR
+{
+  /**
+   * Number type for R matrix.
+   */
+  typedef typename VectorType::value_type Number;
+
+protected:
+  /**
+   * Default private constructor.
+   */
+  BaseQR();
+
+public:
+  /**
+   * Destructor.
+   */
+  virtual ~BaseQR() = default;
+
+  /**
+   * Append @p column to the QR factorization.
+   * Returns <code>true</code> if the result is successful, i.e.
+   * the columns are linearly independent. Otherwise the @p column
+   * is rejected and the return value is <code>false</code>.
+   */
+  virtual bool
+  append_column(const VectorType &column) = 0;
+
+  /**
+   * Remove a column @p k and update QR factorization.
+   */
+  virtual void
+  remove_column(const unsigned int k = 0) = 0;
+
+  /**
+   * Return size of the subspace.
+   */
+  unsigned int
+  size() const;
+
+  /**
+   * Return the current upper triangular matrix R.
+   */
+  const LAPACKFullMatrix<Number> &
+  get_R() const;
+
+  /**
+   * Solve $Rx=y$. Vectors @p x and @p y should be consistent
+   * with the current size of the subspace.
+   * If @p transpose is <code>true</code>, $R^Tx=y$ is solved instead.
+   */
+  void
+  solve(Vector<Number> &      x,
+        const Vector<Number> &y,
+        const bool            transpose = false) const;
+
+  /**
+   * Set $y = Q*x$. The size of $x$ should be consistent with the
+   * size of the R matrix.
+   */
+  virtual void
+  multiply_with_Q(VectorType &y, const Vector<Number> &x) const = 0;
+
+  /**
+   * Set $y = Q^Tx$. The size of $x$ should be consistent with the
+   * size of column vectors.
+   */
+  virtual void
+  multiply_with_QT(Vector<Number> &y, const VectorType &x) const = 0;
+
+  /**
+   * Set $y = Q*R*x$. The size of $x$ should be consistent with the
+   * size of the R matrix.
+   */
+  virtual void
+  multiply_with_A(VectorType &y, const Vector<Number> &x) const = 0;
+
+  /**
+   * Set $y = R^T Q^Tx$. The size of $x$ should be consistent with the
+   * size of column vectors.
+   */
+  virtual void
+  multiply_with_AT(Vector<Number> &y, const VectorType &x) const = 0;
+
+  /**
+   * Connect a slot to retrieve a notification when the Givens rotations
+   * are performed.
+   */
+  boost::signals2::connection
+  connect_givens_slot(
+    const std::function<void(const unsigned int i,
+                             const unsigned int j,
+                             const std::array<Number, 3> &)> &slot);
+
+protected:
+  /**
+   * Multiply with columns stored in the object.
+   */
+  void
+  multiply_with_cols(VectorType &y, const Vector<Number> &x) const;
```

Comment:
What does this mean? That you interpret the columns as the columns of a matrix A and then form y=Ax?

Reply:
it depends on the downstream class. Some store `A`, some `Q`. So the meaning changes, but this is the interface for them to use.

Reply:
I see. Then maybe something like "Compute $y=Hx$ where $H$ is the matrix formed by the column vectors stored by this object."